8359359: AArch64: share trampolines between static calls to the same method

mikabl-arm opened the pull request:
Modify the C2 compiler to share trampoline stubs between static calls that resolve to the same callee method. Since the relocation target for all static calls is initially set to the static call resolver stub, the call's target alone cannot be used to distinguish between different static method calls. Instead, trampoline stubs should be shared based on the actual callee.

The `SharedTrampolineTest.java` was designed to verify the sharing of trampolines among static calls. However, due to imprecise log analysis, the test currently passes even when trampolines are not shared. Additionally, comments within the test suggest ambiguity regarding whether it was intended to assess trampoline sharing for static calls or runtime calls. To address these issues and eliminate ambiguity, this patch renames and updates the existing test. Furthermore, a new test is introduced, using the existing one as a foundation, to accurately evaluate trampoline sharing for both static and runtime calls.

This has passed tier1-3 and jcstress testing on AArch64.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue
- \[ \] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Error
&nbsp;⚠️ Pull request body is missing required line: `- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).`

### Issue
 * [JDK-8359359](https://bugs.openjdk.org/browse/JDK-8359359): AArch64: share trampolines between static calls to the same method (**Enhancement** - P4)


### Reviewers
 * [Evgeny Astigeevich](https://openjdk.org/census#eastigeevich) (@eastig - Committer) 🔄 Re-review required (review applies to [bd065434](https://git.openjdk.org/jdk/pull/25954/files/bd06543472f4fb8f0ee2cf36ac10fe86500d94ae))
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) 🔄 Re-review required (review applies to [bd065434](https://git.openjdk.org/jdk/pull/25954/files/bd06543472f4fb8f0ee2cf36ac10fe86500d94ae))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25954/head:pull/25954` \
`$ git checkout pull/25954`

Update a local copy of the PR: \
`$ git checkout pull/25954` \
`$ git pull https://git.openjdk.org/jdk.git pull/25954/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25954`

View PR using the GUI difftool: \
`$ git pr show -t 25954`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25954.diff">https://git.openjdk.org/jdk/pull/25954.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25954#issuecomment-3001011429)
</details>

mikabl-arm committed:
8359359: AArch64: share trampolines between static calls to the same method

Modify the C2 compiler to share trampoline stubs between static calls
that resolve to the same callee method. Since the relocation target for
all static calls is initially set to the static call resolver stub, the
call's target alone cannot be used to distinguish between different
static method calls. Instead, trampoline stubs should be shared based
on the actual callee.

The `SharedTrampolineTest.java` was designed to verify the sharing of
trampolines among static calls. However, due to imprecise log analysis,
the test currently passes even when trampolines are not shared.
Additionally, comments within the test suggest ambiguity regarding
whether it was intended to assess trampoline sharing for static calls
or runtime calls. To address these issues and eliminate ambiguity, this
patch renames and updates the existing test. Furthermore, a new test is
introduced, using the existing one as a foundation, to accurately
evaluate trampoline sharing for both static and runtime calls.

theRealAph commented:
One thing that looks a little odd: why do you maintain two separate lists and two sets of trampoline stub handlers?

mikabl-arm commented:
@theRealAph , I've tried the above and it seems to work as intended. Please let me know if you find that solution more suitable and I'll prepare a patch for submission.

theRealAph commented:
> We could use something like `Pair<callKind, address>` for keys instead and use the `first` to distinguish between runtime and static calls. I'd need to extend `template<...> class Pair` so it properly supports hashing and comparison (at least equality) for this to work.

Sounds good.

mikabl-arm committed:
Combine the two shared trampoline request hash tables

mikabl-arm committed:
Lift the vm.opt.TieredCompilation == null requirement from the tests

mikabl-arm commented on test/hotspot/jtreg/compiler/sharedstubs/SharedStaticCallTrampolineTest.java:
```diff
@@ -0,0 +1,148 @@
+ * @library /test/lib
+ *
+ * @requires vm.compiler2.enabled
+ * @requires vm.opt.TieredCompilation == null
```
Thank you for a review @lmesnik , fixed!

lmesnik reviewed (commented):
Thanks for fixing tests, please full review from compiler team.

eastig commented on src/hotspot/share/asm/codeBuffer.hpp:
```diff
@@ -541,9 +542,14 @@ class CodeBuffer: public StackObj DEBUG_ONLY(COMMA private Scrubber) {
 
  private:
+  enum class TrampolineCallKind { Runtime, Static };
+  typedef Pair<TrampolineCallKind, address> SharedTrampolineRequestKey;
+  typedef ResizeableResourceHashtable<SharedTrampolineRequestKey, Offsets, AnyObj::C_HEAP,
+                                      mtCompiler>
```
You can keep using `address` as a key. I don't think `ciMethod*` can be a runtime call target. If we keep using `address` as a key we don't need to define a hash function. 
A type of a call can be stored as data with offset.

mikabl-arm replied:
Thank you for a suggestion! I think it works just as fine as the current version does but reduces the footprint of the patch significantly. Hopefully this should make it easier to review the PR. I'll submit an update next week.

mikabl-arm committed:
Address review comments from @theRealAph and @eastig
- use relocInfo::relocType instead of TrampolineCallKind
- move rtype from keys to values; reduce the footprint of the patch

mikabl-arm committed:
Merge branch 'master'

theRealAph commented on src/hotspot/cpu/aarch64/codeBuffer_aarch64.hpp:
```diff
@@ -34,6 +34,15 @@
+ private:
+  void share_trampoline_for(relocInfo::relocType rtype, address dest, int caller_offset);
+
+ public:
+  void share_rc_trampoline_for(address dest, int caller_offset) {
+    share_trampoline_for(relocInfo::runtime_call_type, dest, caller_offset);
+  }
+  void share_sc_trampoline_for(const ciMethod *callee, int caller_offset) {
+    share_trampoline_for(relocInfo::static_call_type, (address)callee, caller_offset);
+  }
```
```suggestion
 public:
  void share_runtime_trampoline_for(address dest, int caller_offset) {
    share_trampoline_for(relocInfo::runtime_call_type, dest, caller_offset);
  }
  void share_static_trampoline_for(const ciMethod *callee, int caller_offset) {
    share_trampoline_for(relocInfo::static_call_type, (address)callee, caller_offset);
  }
```

eastig commented on src/hotspot/share/asm/codeBuffer.hpp:
```diff
@@ -540,7 +541,7 @@ class CodeBuffer: public StackObj DEBUG_ONLY(COMMA private Scrubber) {
   };
 
-  typedef LinkedListImpl<int> Offsets;
+  typedef Pair<relocInfo::relocType, LinkedListImpl<int>> Offsets;
   typedef ResizeableResourceHashtable<address, Offsets, AnyObj::C_HEAP, mtCompiler> SharedTrampolineRequests;
```
I think code can be simplified.
For example, let's a Java method `A` call a static Java method `B` twice and a static Java method `C` three times.
You will be having the following in `SharedTrampilneRequests`:
```
key -> value
{B} -> {relocInfo::static_call_type, {call_B_offset1, call_B_offset2}}
{C} -> {relocInfo::static_call_type, {call_C_offset1, call_C_offset2, call_C_offset3}}
```

When you iterate trampoline requests, you don't use the key for a destination if it is for a static call. Instead you use `SharedRuntime::get_resolve_static_call_stub()` as a destination. 

This means your requests don't differ from requests for `runtime_call_type` trampolines. `runtime_call_type` trampolines have final destinations but `static_call_type` have a resolver as a destination.

So you actually have:
```
key -> value
(trampoline_id1} -> {get_resolve_static_call_stub, {call_B_offset1, call_B_offset2}}
{trampoline_id2} -> {get_resolve_static_call_stub, {call_C_offset1, call_C_offset2, call_C_offset3}}
{trampoilen_id3} -> {some_runtime_call, {call_RT_offset1, call_RT_offset2, call_RT_offset3}}
```

IMO based on `SharedRuntime::resolve_helper`,  we can have shared trampolines for `relocInfo::opt_virtual_call_type` and `relocInfo::virtual_call_type`. Their trampolines have the corresponding resolver set. I don't see we write anything specific to a call site into a trampoline.

We can have the following:
```
class SharedTrampolineRequest {
   LinkedListImpl<int> _offsets;
   address _dest;
#ifndef PRODUCT
   relocInfo::relocType; // This can be used in asserts to check a shared trampoline is for the same relocType
#endif
};

// A set of requests for shared trampolines.
// We use a hash table to implement the set.
// A trampoline id is a key in the table.
// We don't store an id in SharedTrampolineRequest because it is only used for the table.
typedef uintptr_t SharedTrampolineId;
typedef ResizeableResourceHashtable<SharedTrampolineId, SharedTrampolineRequest,
     AnyObj::C_HEAP, mtCompiler> SharedTrampolineRequests;
```

You will need to modify `MacroAssembler::trampoline_call`:

```
SharedTrampolineId tramp_id = (entry.rspec().type() == relocInfo::runtime_call_type) ? target : callee;
code()->share_trampoline_for(tramp_id, entry.target(), offset());
```

New code for `share_trampoline_for(NOT_PRODUCT_ARG(relocInfo::relocType rel_type) tramp_id, target, offset )`:
```
bool created;
SharedTrampolineRequest tramp_req = _shared_trampoline_requests->put_if_absent(tramp_id, &created);
  if (created) {
    _shared_trampoline_requests->maybe_grow();
    tramp_req->set_target(target);
    NOT_PRODUCT(tramp_req->set_reloc_type(rel_type));
  }
  assert(tramp_req->reloc_type() == rel_type);
  tramp_req->offsets()->add(offset);
  _finalize_stubs = true;
```

eastig replied:
> Am I right assuming you're suggesting to simplify ...

Yes.

> However, I'm not sure there's a clearly better approach here. This design leads to some duplication - for instance, trampoline_id3 should match some_runtime_call, if I understand correctly.

There should not be any duplication. Why do you think so?

Let's revisit the example. This time `A` has three calls of  `Arrays.hashCode`. For large arrays we will use the stub `StubRoutines::aarch64::large_arrays_hashcode`.

Your implementation:
```
key -> value
{B} -> {relocInfo::static_call_type, {call_B_offset1, call_B_offset2}}
{C} -> {relocInfo::static_call_type, {call_C_offset1, call_C_offset2, call_C_offset3}}
{StubRoutines::aarch64::large_arrays_hashcode} -> {relocInfo::runtime_call_type, {call_RT_offset1, call_RT_offset2, call_RT_offset3}}
```

Mine proposal:
```
key -> value
(B} -> {get_resolve_static_call_stub, {call_B_offset1, call_B_offset2}}
{C} -> {get_resolve_static_call_stub, {call_C_offset1, call_C_offset2, call_C_offset3}}
{StubRoutines::aarch64::large_arrays_hashcode} -> {StubRoutines::aarch64::large_arrays_hashcode, {call_RT_offset1, call_RT_offset2, call_RT_offset3}}
```

I don't see any duplication.

> Taking this into account, I'd suggest keeping the current approach and not proceeding with the alternative implementation described above.

I dont understand what you mean "the current approach". Do you mean using `Pair` and checking `relocInfo::relocType`?

`Pair` obfuscates what is stored inside it. IMO this reduces readability of the program. `SharedTrampolineRequest` clearly states its purpose and the purpose of its members.

> I can see this being implemented as a follow-up PR.
I am ok with this.

mikabl-arm replied:
> There should not be any duplication. Why do you think so?

`StubRoutines::aarch64::large_arrays_hashcode` is both a key and a part of the value in the example above.

> I dont understand what you mean "the current approach". Do you mean using Pair and checking relocInfo::relocType?

Mostly the latter. I agree that the implementation would benefit from using something more explicit instead of `Pair<>`.

eastig reviewed (changes requested):
This PR needs additional research and changes.

mikabl-arm committed:
Remove share_rc_trampoline_for and share_sc_trampoline_for

mikabl-arm committed:
Make the value type of the dictionary a struct instead of Pair typedef

mikabl-arm committed:
cleanup: update copyright headers

mikabl-arm committed:
remove implementation-dependent logic from emit_shared_trampolines()

mikabl-arm committed:
Merge commit '41520998aa8808452ee384b213b2a77c7bad668d'

mikabl-arm commented on src/hotspot/cpu/aarch64/codeBuffer_aarch64.cpp:
```diff
@@ -50,10 +56,14 @@ static bool emit_shared_trampolines(CodeBuffer* cb, CodeBuffer::SharedTrampoline
     int offset = *it.next();
+    if (rtype == relocInfo::static_call_type) {
+      dest = SharedRuntime::get_resolve_static_call_stub();
+    }
```
Hi @eastig , I hope you find this implementation more appropriate, please check https://github.com/openjdk/jdk/pull/25954/commits/d03d8384570b22141f81f06249a44387b89c8303

eastig commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -905,7 +906,10 @@ address MacroAssembler::trampoline_call(Address entry) {
         assert(CodeBuffer::supports_shared_stubs(), "must support shared stubs");
-        code()->share_trampoline_for(entry.target(), offset());
+        code()->share_trampoline_for(entry.target(), entry.target(), offset());
+      } else if (entry.rspec().type() == relocInfo::static_call_type && callee != nullptr) {
```
I think we should require `callee` to non-null for `relocInfo::static_call_type`. Otherwise this will hide bugs, when it's been intended to share but forgotten to provide `callee`.

eastig replied:
We need to convert the check `callee != nullptr` into an assert.

dean-long replied:
Are you sure this is OK in LIR_Assembler::call()?  Calls of type lir_dynamic_call use relocInfo::static_call_type, but can't they resolve to different targets depending on the call site info?  I don't think op->method() is unique for invokedynamic calls.  @iwanowww what do you think?

mikabl-arm replied:
> Are you sure this is OK in LIR_Assembler::call()?

Fixed, please see https://github.com/openjdk/jdk/pull/25954/commits/ba317d4ca96af87fbc53155d1fcc33d88a6c2349

eastig commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:
```diff
@@ -1333,10 +1334,14 @@ class MacroAssembler: public Assembler {
   // - relocInfo::virtual_call_type
+  // Trampolines may be emitted immediately or deferred until stub finalization,
+  // enabling reuse across call sites to reduce code size.
+  // Runtime call trampolines are shared based on the entry value.
```
... on entry value -> ... on entry target

mikabl-arm committed:
Merge commit 'fd29677479797956e0d205b5ce6e7cb9ad407bd1' into 8359359

mikabl-arm committed:
cleanup: rephrase comments in macroAssembler_aarch64.hpp

mikabl-arm committed:
assert that call sites offsets aren't missing

mikabl-arm committed:
assert callee is nullptr for runtime calls

mikabl-arm committed:
share static call trampolines generated by C1 as well

mikabl-arm committed:
fixup: a shared trampoline must branch to a statically bound method

mikabl-arm committed:
the only trampoline in ArrayCopyStub is never shared

mikabl-arm committed:
Merge commit 'f6f87bb6759c86d941453a1776e8abfdffc48183' into 8359359

mikabl-arm commented on src/hotspot/cpu/aarch64/c1_CodeStubs_aarch64.cpp:
```diff
@@ -332,7 +333,7 @@ void ArrayCopyStub::emit_code(LIR_Assembler* ce) {
   Address resolve(SharedRuntime::get_resolve_static_call_stub(),
                   relocInfo::static_call_type);
-  address call = __ trampoline_call(resolve);
+  address call = __ trampoline_call(resolve, info()->method());
```
reverted, please see https://github.com/openjdk/jdk/pull/25954/commits/a5c665520088ca7a3f282b684bb58701733af83e

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedRuntimeCallTrampolineTest.java:
```diff
@@ -72,38 +73,64 @@ private static void runTest(String compiler, String test) throws Exception {
+                .filter(reloc -> callAddrs.contains(reloc.owner()))
+                .map(reloc -> new String(reloc.addr()))
+                .collect(Collectors.toList());
+        if (trampolineAddrs.stream().distinct().count() >= trampolineAddrs.size()) {
```
For better readability, could you please create a meaningful variable for `trampolineAddrs.stream().distinct().count()`? You can reuse it in the exception message as well.

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedStaticCallTrampolineTest.java:
```diff
@@ -0,0 +1,147 @@
+        List<String> uniqueTrampolineAddrs = trampolineAddrs.stream()
+                .filter(addr -> Collections.frequency(trampolineAddrs, addr) == 1)
+                .collect(Collectors.toList());
+        if (uniqueTrampolineAddrs.size() == 0) {
```
Should we expect this to be 1? Possible values: 0, 1 or 3?
0 - incorrect mapping of a call site
1 - everything is correct
3 - sharing does not work

mikabl-arm replied:
Yes, the code expects this to be 1. I'd argue that testing for 3 here would duplicate the check already performed above ([line 119](https://github.com/openjdk/jdk/pull/25954/files#diff-96cb53c21e0d1344a4359e395671d20c5c4dc8493a87c7716936f35258f939cfR119)).

Currently, `checkOutput(OutputAnalyzer)` explicitly checks for two kinds of incorrect behaviors:

1. No trampoline stubs shared across static calls
2. A trampoline stub is shared while it should not be

If neither of these issues is detected, the code implicitly treats the behavior as correct.

mikabl-arm committed:
cleanup: make tests source code more readable

mikabl-arm committed:
Merge commit '8bafc2f0aecbbe548573712a9dc31c9764f82f71' into 8359359

eastig reviewed (approved):
lgtm

dean-long commented:
A question about copyright headers.  This PR was posted in 2025, so should copyright dates remain 2025 or get updated to 2026?  I think Oracle scripts expect copyright year to match commit date.  Also, I believe all openjdk files need an Oracle copyright header line, but I don't see one in codeBuffer_aarch64.cpp.

dean-long reviewed (approved):
This looks good.  Also I ran testing and it passed.

mikabl-arm committed:
cleanup: update copyright notices to 2026

mikabl-arm committed:
cleanup: list variables captured by a lambda explicitly

mikabl-arm committed:
cleanup: do note replace a year in a copyright notice, use a range

mikabl-arm committed:
cleanup: use a comma to denote a range in a copyright notice

mikabl-arm commented:
/touch

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedRuntimeCallTrampolineTest.java:
```diff
@@ -0,0 +1,145 @@
+ */
+
+/**
+ * @test SharedRuntimeCallTrampolineTest id=C2
```
Can we have a test for C1 as well?

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedStaticCallTrampolineTest.java:
```diff
@@ -0,0 +1,153 @@
+ */
+
+/**
+ * @test SharedStaticCallTrampolineTest id=C2
```
Can we have a test for C1 as well?

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedStaticCallTrampolineTest.java:
```diff
@@ -0,0 +1,153 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
```
Unused import

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedStaticCallTrampolineTest.java:
```diff
@@ -0,0 +1,153 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
```
Unused import

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedStaticCallTrampolineTest.java:
```diff
@@ -0,0 +1,153 @@
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.regex.MatchResult;
```
Unused imports

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedRuntimeCallTrampolineTest.java:
```diff
@@ -0,0 +1,145 @@
+package compiler.sharedstubs;
+
+import java.util.ArrayList;
+import java.util.Collections;
```
Unused import

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedRuntimeCallTrampolineTest.java:
```diff
@@ -0,0 +1,145 @@
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.regex.Pattern;
+import jdk.test.lib.process.OutputAnalyzer;
```
Missing a blank line between java.* import group and jdk.* import group. Also imports inside each group should be sorted alphabetically. Similar comment for another test.

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedRuntimeCallTrampolineTest.java:
```diff
@@ -0,0 +1,145 @@
+
+        List<String> trampolineAddrs = trampolineRelocs.stream()
+                .filter(reloc -> callAddrs.contains(reloc.owner()))
+                .map(reloc -> new String(reloc.addr()))
```
`new String()` is redundant. `reloc.addr()` is already a String.

eastig commented on test/hotspot/jtreg/compiler/sharedstubs/SharedStaticCallTrampolineTest.java:
```diff
@@ -0,0 +1,153 @@
+
+        List<String> trampolineAddrs = trampolineRelocs.stream()
+                .filter(reloc -> callAddrs.contains(reloc.owner()))
+                .map(reloc -> new String(reloc.addr()))
```
`new String()` is redundant. `reloc.addr()` is already a String.

eastig commented:
@mikabl-arm 
The following in the description is not correct any more:

> Modify the C2 compiler to share trampoline stubs between static calls that resolve to the same callee method.
